File: app/src/main/java/me/ykrank/s1next/view/fragment/ForumFragment.java
package me.ykrank.s1next.view.fragment;

import android.app.Activity;
import android.content.Context;
import android.net.Uri;
import android.os.Bundle;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;

import me.ykrank.s1next.R;
import me.ykrank.s1next.data.api.Api;
import me.ykrank.s1next.data.api.model.collection.ForumGroups;
import me.ykrank.s1next.data.api.model.wrapper.ForumGroupsWrapper;
import me.ykrank.s1next.util.IntentUtil;
import me.ykrank.s1next.view.activity.SearchActivity;
import me.ykrank.s1next.view.adapter.ForumRecyclerViewAdapter;
import me.ykrank.s1next.view.internal.ToolbarDropDownInterface;
import rx.Observable;

/**
 * A Fragment represents forum list.
 */
public final class ForumFragment extends BaseRecyclerViewFragment<ForumGroupsWrapper>
        implements ToolbarDropDownInterface.OnItemSelectedListener {

    public static final String TAG = ForumFragment.class.getName();

    private ForumRecyclerViewAdapter mRecyclerAdapter;
    private ForumGroups mForumGroups;

    private ToolbarDropDownInterface.Callback mToolbarCallback;

    @Override
    public void onViewCreated(View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        RecyclerView recyclerView = getRecyclerView();
        recyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
        mRecyclerAdapter = new ForumRecyclerViewAdapter(getActivity());
        recyclerView.setAdapter(mRecyclerAdapter);
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);

        mToolbarCallback = (ToolbarDropDownInterface.Callback) context;
    }

    @Override
    public void onDetach() {
        super.onDetach();

        mToolbarCallback = null;
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        super.onCreateOptionsMenu(menu, inflater);
        inflater.inflate(R.menu.fragment_forum, menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.menu_browser:
                IntentUtil.startViewIntentExcludeOurApp(getContext(), Uri.parse(Api.BASE_URL));

                return true;
            case R.id.app_bar_search:
                Activity activity = getActivity();
                SearchActivity.start(activity, activity.findViewById(R.id.app_bar_search));
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    @Override
    Observable<ForumGroupsWrapper> getSourceObservable() {
        return mS1Service.getForumGroupsWrapper();
    }

    @Override
    void onNext(ForumGroupsWrapper data) {
        super.onNext(data);

        mForumGroups = data.getForumGroups();
        // host activity would call #onToolbarDropDownItemSelected(int) after
        mToolbarCallback.setupToolbarDropDown(mForumGroups.getForumGroupNameList());
    }

    /**
     * Show all forums when {@code position == 0} otherwise show
     * corresponding forum group's forum list.
     */
    @Override
    public void onToolbarDropDownItemSelected(int position) {
        if (position == 0) {
            mRecyclerAdapter.refreshDataSet(mForumGroups.getForumList(), true);
        } else {
            // the first position is "全部"
            // so position - 1 to correspond its group
            mRecyclerAdapter.refreshDataSet(mForumGroups.getForumGroupList().get(position - 1), true);
        }
    }
}

File: app/src/main/java/me/ykrank/s1next/view/fragment/ForumFragment.kt
package me.ykrank.s1next.view.fragment

import android.content.Context
import android.net.Uri
import android.os.Bundle
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import androidx.lifecycle.lifecycleScope
import com.github.ykrank.androidtools.data.CacheParam
import com.github.ykrank.androidtools.data.Resource
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.launch
import me.ykrank.s1next.App
import me.ykrank.s1next.R
import me.ykrank.s1next.data.api.Api
import me.ykrank.s1next.data.api.model.collection.ForumGroups
import me.ykrank.s1next.data.api.model.wrapper.ForumGroupsWrapper
import me.ykrank.s1next.util.IntentUtil
import me.ykrank.s1next.view.activity.SearchActivity
import me.ykrank.s1next.view.adapter.ForumRecyclerViewAdapter
import me.ykrank.s1next.view.event.LoginEvent
import me.ykrank.s1next.view.internal.ToolbarDropDownInterface

/**
 * A Fragment represents forum list.
 */
class ForumFragment : BaseRecyclerViewFragment<ForumGroupsWrapper>(),
    ToolbarDropDownInterface.OnItemSelectedListener {
    private lateinit var mRecyclerAdapter: ForumRecyclerViewAdapter
    private var mForumGroups: ForumGroups? = null

    private var mToolbarCallback: ToolbarDropDownInterface.Callback? = null

    private var inForceRefresh = false

    override fun onAttach(context: Context) {
        App.appComponent.inject(this)
        super.onAttach(context)

        mToolbarCallback = context as ToolbarDropDownInterface.Callback?
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        leavePageMsg("ForumFragment")

        val recyclerView = recyclerView
        recyclerView.layoutManager = androidx.recyclerview.widget.LinearLayoutManager(context)
        mRecyclerAdapter = ForumRecyclerViewAdapter(activity)
        recyclerView.adapter = mRecyclerAdapter

        lifecycleScope.launch {
            mEventBus.getClsFlow<LoginEvent>()
                .collect {
                    forceSwipeRefresh()
                }
        }
    }

    override fun onDetach() {
        super.onDetach()

        mToolbarCallback = null
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        super.onCreateOptionsMenu(menu, inflater)
        inflater.inflate(R.menu.fragment_forum, menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            R.id.menu_browser -> {
                IntentUtil.startViewIntentExcludeOurApp(requireContext(), Uri.parse(Api.BASE_URL))
                true
            }

            R.id.app_bar_search -> {
                val activity = requireActivity()
                SearchActivity.start(activity, activity.findViewById(R.id.app_bar_search))
                true
            }

            else -> super.onOptionsItemSelected(item)
        }
    }

    override suspend fun getSource(loading: Int): Flow<Resource<ForumGroupsWrapper>> {
        return apiCacheProvider.getForumGroupsWrapper(CacheParam(isForceLoading))
    }

    override fun onNext(data: ForumGroupsWrapper) {
        super.onNext(data)

        mForumGroups = data.data
        // host activity would call #onToolbarDropDownItemSelected(int) after
        mToolbarCallback?.setupToolbarDropDown(mForumGroups?.forumGroupNameList ?: emptyList())
    }

    /**
     * Show all forums when `position == 0` otherwise show
     * corresponding forum group's forum list.
     */
    override fun onToolbarDropDownItemSelected(position: Int) {
        mForumGroups?.let {
            if (position == 0) {
                mRecyclerAdapter.refreshDataSet(it.forumList, true)
            } else {
                // the first position is "全部"
                // so position - 1 to correspond its group
                mRecyclerAdapter.refreshDataSet(
                    it.forumGroupList.getOrNull(position - 1) ?: emptyList(), true
                )
            }
        }
        if (inForceRefresh) {
            inForceRefresh = false

            recyclerView.post {
                recyclerView.smoothScrollToPosition(0)
            }
        }
    }

    private fun forceSwipeRefresh() {
        inForceRefresh = true
        startSwipeRefresh()
    }

    companion object {
        val TAG = ForumFragment::class.java.simpleName
    }
}
